### PR TITLE
Serialization to memory

### PR DESCRIPTION
**Pull request by likholat**

### CODE REVIEW

File: inference-engine/tests/functional/inference_engine/ir_serialization/deterministicity.cpp [hunk 1]
```diff
@@ -124,3 +124,21 @@ TEST_F(SerializationDeterministicityTest, ModelWithConstants) {
...
+    InferenceEngine::Blob::Ptr m_out_bin_buf;
+
+    InferenceEngine::Core ie;
+    auto expected = ie.ReadNetwork(model, weights);
+    expected.serialize(m_out_xml_buf, m_out_bin_buf);
+    auto result = ie.ReadNetwork(m_out_xml_buf.str(), m_out_bin_buf);
+
+    ASSERT_TRUE(expected.layerCount() == result.layerCount());
+    ASSERT_TRUE(expected.getInputShapes() == result.getInputShapes());
+}
```
Comment (dkurt) [inline inference-engine/tests/functional/inference_engine/ir_serialization/deterministicity.cpp:144]: Add one more test for two streams

**Comment by ilya-lavrenov** [comment]
why do we need it?

**Comment by dkurt** [comment]
@ilya-lavrenov, this is a missed functionality. There is customer request for it. Usecase is network serialization in runtime with encryption.

**Comment by ilya-lavrenov** [comment]
Why customer needs to serialize network in runtime? What is the usecase?

**Comment by dkurt** [comment]
@ilya-lavrenov, to have an approach to encrypt model as a binary buffer and transfer between different OV sessions without dumping IR. Initially models are built in runtime with nGraph without MO.

**Review by dkurt — APPROVED** [review]

File: inference-engine/include/cpp/ie_cnn_network.h [hunk 2]
```diff
@@ -199,6 +199,22 @@ class INFERENCE_ENGINE_API_CLASS(CNNNetwork) {
...
+     */
+    void serialize(std::ostream& xmlBuf, std::ostream& binBuf) const;
+
+    /**
+     * @brief Serialize network to IR stream and weights Blob::Ptr.
+     *
+     * @param xmlBuf output IR stream.
+     * @param binBlob output weights Blob::Ptr.
+     */
+    void serialize(std::ostream& xmlBuf, Blob::Ptr& binBlob) const;
```
Comment (ilya-lavrenov) [inline inference-engine/include/cpp/ie_cnn_network.h:216]: Do we need both overloads ?

Comment (ilyachur) [reply inference-engine/include/cpp/ie_cnn_network.h:216]: Why do we need serialization to blob?

Comment (dkurt) [reply inference-engine/include/cpp/ie_cnn_network.h:216]: Because we have corresponding ReadNetwork with Blob

**Review by ilya-lavrenov — APPROVED** [review]

File: inference-engine/src/inference_engine/cnn_network_ngraph_impl.cpp [hunk 3]
```diff
@@ -476,6 +480,64 @@ StatusCode CNNNetworkNGraphImpl::serialize(const std::string& xmlPath,
...
+            custom_opsets);
+        manager.run_passes(_ngraph_function);
+
+        std::streambuf* pbuf = binBuf.rdbuf();
+        unsigned long bufSize = binBuf.tellp();
+
+        TensorDesc tensorDesc(Precision::U8, { bufSize }, Layout::C);
+        binBlob = make_shared_blob<uint8_t>(tensorDesc);
+        binBlob->allocate();
+        pbuf->sgetn(binBlob->buffer(), bufSize);
```
Comment (ilya-lavrenov) [inline inference-engine/src/inference_engine/cnn_network_ngraph_impl.cpp:530]: Constant data doubles memory consumption here, but hard to imagine how to overcome this..

**Review by ilyachur — APPROVED** [review]

**Pull request merged**